### PR TITLE
Language 2

### PR DESCRIPTION
**Pull request by jsinglet**
## Description

Implements the language 2 package. Note this also provides refactorings for incorrectly named rules.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - RULE-1-4
     - RULE-4-2
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/common/test/rules/usageofassemblernotdocumented/test.c [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// COMPLIANT
```
Comment (knewbury01) [inline c/common/test/rules/usageofassemblernotdocumented/test.c:1]: er sorry I asked this before but I forget the answer, was there any way to avoid having the exact same file 2x, like some way to ref the testcase from the other shared implementation? not a big deal if not!

**Review by knewbury01 — COMMENTED** [review]
DIR-4-2 reviewed - shared query formatting double checked

rule->dir namings verified

everything looks really good so far! most minor comments ever, keep up the good work! :)

File: scripts/util/Get-DuplicateRules.ps1 [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env pwsh
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply scripts/util/Get-DuplicateRules.ps1:1]: lol, but does your one liner have two different CI modes and accounting for the number of duplicated rules? :)

[the remark this answers is not in the document]
Comment (knewbury01) [reply scripts/util/Get-DuplicateRules.ps1]: haha it does not! XD this truly is a case of one script to rule them all!

**Review by knewbury01 — COMMENTED** [review]
@jsinglet RULE-1-4- reviewed, I like the use of modules and abstract classes, tidy implementation! nice work :) just had a few minor questions!

File: rule_packages/c/Language2.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      "properties": {
+        "obligation": "advisory"
+      },
+      "queries": [
+        {
+          "description": "",
+          "kind": "problem",
+          "name": "All usage of assembly language should be documented",
+          "precision": "very-high",
+          "severity": "error",
```
Comment (knewbury01) [inline rule_packages/c/Language2.json:13]: should these (this and RULE-1-4) be warning? they arent necessarily undefined behaviour level, just unpredicatable?

File: rule_packages/c/Language2.json [hunk 4]
```diff
@@ -0,0 +1,45 @@
+{
+  "MISRA-C-2012": {
+    "DIR-4-2": {
+      "properties": {
+        "obligation": "advisory"
+      },
+      "queries": [
+        {
+          "description": "",
```
Comment (knewbury01) [inline rule_packages/c/Language2.json:9]: also missing description

File: c/misra/test/rules/RULE-1-4/test.c [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#define MACRO(x) _Generic((x), int : 0, long : 1) // NON_COMPLIANT
+#define __STDC_WANT_LIB_EXT1__ 1                  // NON_COMPLIANT
+
+_Noreturn void f0(); // NON_COMPLIANT
+
+typedef int new_type;                     // COMPLIANT
+typedef _Atomic new_type atomic_new_type; // NON_COMPLIANT
+
+void f(int p) {
+  // int i0 = _Generic(p, int : 0, long : 1); // NON_COMPLIANT[FALSE_NEGATIVE]
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply c/misra/test/rules/RULE-1-4/test.c:15]: It doesn't compile

File: cpp/common/src/codingstandards/cpp/Emergent.qll [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      ThreadLocalDeclaration() { getASpecifier().getName() = "is_thread_local" }
+    }
+  
+    class EmergentHeader extends EmergentLanguageFeature, Include {
+      EmergentHeader() {
+        getIncludedFile().getBaseName() = ["stdalign.h", "stdatomic.h", "stdnoreturn.h", "threads.h"]
+      }
+    }
+  
+    class LibExt1Macro extends EmergentLanguageFeature, Macro {
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply cpp/common/src/codingstandards/cpp/Emergent.qll:33]: Good point -- changed!

File: c/misra/test/rules/RULE-1-4/test.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/test/rules/RULE-1-4/test.c:15]: erp, which clang did you use?

File: cpp/common/src/codingstandards/cpp/Emergent.qll (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsinglet) [reply cpp/common/src/codingstandards/cpp/Emergent.qll:33]: After rereading the spec, I've reversed on this. It explicitly states 1 and 0 are the only permitted values.

File: rule_packages/c/Language2.json (hunk 3 above)
Comment (jsinglet) [reply rule_packages/c/Language2.json:13]: Agree!

File: rule_packages/c/Language2.json (hunk 4 above)
Comment (jsinglet) [reply rule_packages/c/Language2.json]: Thanks!

File: rule_packages/c/Language2.json [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+          "shared_implementation_short_name": "UsageOfAssemblerNotDocumented",
+          "tags": [
+            "maintainability",
+            "readability" 
+          ]
+        }
+      ],
+      "title": "All usage of assembly language should be documented"
+    },
+    "RULE-1-4": {
```
[the remark this answers is not in the document]
Comment (jsinglet) [reply rule_packages/c/Language2.json:24]:
```1 The functions, macros, and types declared or defined in K.3 and its subclauses are not declared or defined by their respective headers if __STDC_WANT_LIB_EXT1__ is defined as a macro which expands to the integer constant 0 at the point in the source file where the appropriate header is first included.

2 The functions, macros, and types declared or defined in K.3 and its subclauses are declared and defined by their respective headers if __STDC_WANT_LIB_EXT1__ is defined as a macro which expands to the integer constant 1 at the point in the source file where the appropriate header is first included.381)

3 It is implementation-defined whether the functions, macros, and types declared or defined in K.3 and its subclauses are declared or defined by their respective headers if __STDC_WANT_LIB_EXT1__ is not defined as a macro at the point in the source file where the appropriate header is first included.382)

4 Within a preprocessing translation unit, __STDC_WANT_LIB_EXT1__ shall be defined identically for all inclusions of any headers from subclause K.3. If __STDC_WANT_LIB_EXT1__ is defined differently for any such inclusion, the implementation shall issue a diagnostic as if a preprocessor error directive were used.```

Case 1: __STDC_WANT_LIB_EXT1__  = 0, DO NOT declare
Case 2: __STDC_WANT_LIB_EXT1__ = 1, DECLARE
Case 3: If it's missing, it is undefined if it is there at that point in the processing.

but Case 1 prohibits it from being used unless __STDC_WANT_LIB_EXT1__  is set to 1 somewhere so I think this is sufficient -- that is for us not to be looking for abnormal compiler behavior.

[the remark this answers is not in the document]
Comment (jsinglet) [reply rule_packages/c/Language2.json:24]: It's worth noting that these functions are generally not supported by clang and thus cannot be easily tested beyond the macro. Example:

```
#if defined(__STDC_LIB_EXT1__)
#define  __STDC_WANT_LIB_EXT1__ 1
#include <stdio.h>
#else
#error "__STDC_LIB_EXT1__ not present"
#endif

int main(int argc, char** argv)
{
    char t[l]; 
    tmpnam_s(t, l);
    return 0;
}
```

does not compile under clang. 

```

[the remark this answers is not in the document]
Comment (knewbury01) [reply rule_packages/c/Language2.json:24]: sorry to keep picking on this case, but is this guard meant to be `#if !defined(__STDC_LIB_EXT1__)`, isnt the guard logic usually like, if not yet defined - define?

File: rule_packages/c/Language2.json (hunk 3 above)
Comment (knewbury01) [reply rule_packages/c/Language2.json:13]: I think you missed this one, still is error

File: c/misra/test/rules/RULE-1-4/test.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/test/rules/RULE-1-4/test.c:15]: verified clang10!

[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/test/rules/RULE-1-4/test.c:15]: erg, so brew doesnt have clang 10 (and I only have apple clang 14 currently), so I am going to make you reverify - 
for me with:
```
Apple clang version 14.0.0 (clang-1400.0.29.102)
Target: arm64-apple-darwin21.6.0
Thread model: posix
```
this compiles
```
#if !defined(__STDC_LIB_EXT1__)
#define  __STDC_WANT_LIB_EXT1__ 1
#include <errno.h>
#else
#error "__STDC_LIB_EXT1__ not present"
#endif

int main(int argc, char** argv)
{
  int p = 0;
  int i0 = _Generic(p, int : 0, long : 1);
  errno_t i = 0;
  return i;
}
```

**Review by knewbury01 — COMMENTED** [review]
@jsinglet had another look! after like 2 more small confirmations I think will be ready! thanks again for all the compiler back n forth!

File: c/misra/test/rules/RULE-1-4/test.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsinglet) [reply c/misra/test/rules/RULE-1-4/test.c:15]: Sorry, I mispoke @knewbury01 -- the issue wasn't that it doesn't compile, the issue is that it is not in the AST. Verified with C team.

File: rule_packages/c/Language2.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (knewbury01) [reply rule_packages/c/Language2.json:24]: not relevant for the testcases anyhow, and reverified that the cases in question do not compile with clang 10! all good!

**Review by knewbury01 — APPROVED** [review]
LGTM now, good work!
